File: Algo/Lab1/main.py
from math import log2

# 1. Реализовать функцию, которая на вход получает натуральное четное число и возвращает два простых числа, 
# которые в сумме составляют входное число (с подсчетом количества итераций).
# 2. Написать функцию, которая оценивает трудоемкость алгоритма из п. 1 для следующих значений: 
# 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096.

def numToSimples(inp):
    itr = 0
    for i in range(1, inp):
        simp, itr = isSimple(i, itr)
        if simp:
            simp, itr = isSimple(inp - i, itr)
            if simp:
                return itr

def isSimple(num, iterations):
    for i in range(2, num // 2 + 1):
        iterations += 1
        if num % i == 0:
            return False, iterations
    return True, iterations

def main():
    tests = [8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096]
    print("num     | Te1    | T1     | Te1/T1 ")
    print('-' * 35)
    for i in tests:
        print(f'{i:<8}|', end='')
        test = numToSimples(i)
        print(f' {test:<7}|', end='')
        print(f' {(i)//2:<7}|', end='')
        print(f' {test/((i)//2)}')


if __name__ == '__main__':
    main()

File: Algo/Lab1/main.cpp
#include <iostream>
#include <vector>

bool isSimple(int num, int &iterations) {
    for (int i = 2; i <= num / 2; ++i) {
        ++iterations;
        if (num % i == 0) {
            return false;
        }
    }
    return true;
}

int numToSimples(int inp) {
    int itr = 0;
    for (int i = 1; i < inp / 2; ++i) {
        bool simp = isSimple(i, itr);
        if (simp) {
            simp = isSimple(inp - i, itr);
            if (simp) {
                return itr;
            }
        }
    }
    return itr;
}

int main() {
    std::vector<int> tests = {8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096};
    std::cout << "num     | Te1    | T1     | Te1/T1 " << std::endl;
    std::cout << std::string(35, '-') << std::endl;
    
    for (int i : tests) {
        std::cout << i << "\t| ";
        int test = numToSimples(i);
        std::cout << test << "\t| ";
        std::cout << (i / 2) << "\t| ";
        std::cout << (static_cast<double>(test) / (i / 2)) << std::endl;
    }
    return 0;
}
